Remove pluralization of email subscription topic prefixes

ChrisBAshton opened the pull request:
Trello: https://trello.com/c/l9XfLavV/3212-email-alerts-on-edit-finder-form

Support for a hash value for `subscription_list_title_prefix` has been around for ages (at least [9 years](b2e3d30#diff-793b8db55b8b5c75a5bf7007ed820f5038b1e7f6160a7f43a12477372fc3c3ceR16)) but it is only used by a handful of finders.

A [whole bunch of logic](https://github.com/alphagov/finder-frontend/blob/484396b660715811500af3e3b226ecf0dd69c675/app/lib/email_alert_title_builder.rb#L19-L47) has to be maintained in finder-frontend to support plurality. Until now, we haven't felt the pain of it in Specialist Publisher, but we're now building a form whereby publishers can request edits to the configuration of their finder. That means we either need to add support for pluralised subscription topics in the form itself, or skip over any existing hash values, or, as we're doing in this commit, omit the hash values altogether. This commit argues that the feature is poorly utilised, not very valuable, and is actively blocking development of the "request edits to email alerts" form.

Only 6 finders made use of the feature, and for five of these, we were able to trivially replace references to "...the following {nouns|noun}" to "...the following noun(s)". There is one finder that had "category" vs "categories", which is harder to substitute, but [taking inspiration from other finders](https://github.com/alphagov/specialist-publisher/blob/b638d1326b570988d63aaec9de2bd6c2c877ea22/lib/documents/schemas/flood_and_coastal_erosion_risk_management_research_reports.json#L17) we can simply follow the form "{Finder name} {document noun}", with no need to overcomplicate things by trying to form a partial sentence.

With this commit, we can delete the corresponding code from finder-frontend.

NB, I wasn't sure what would happen to existing subscriber lists if we were to go and change the subscriber list title (but keeping all other params the same) - is a new one created, and if so, is that even a problem? So I tried it out. Rails console on Integration (Email Alert API), getting a subscriber list:

``` email-alert-api(prod)> SubscriberList.find_by(slug: "upper-tribunal-tax-and-chancery-chamber-decisions-with-the-following-category-charity") => id: 8330, title: "Upper Tribunal (Tax and Chancery Chamber) decision...", created_at: "2017-04-04 11:25:58.190324000 +0100", updated_at: "2019-09-16 13:26:39.390801000 +0100", document_type: "", tags: {"format"=>{"any"=>["tax_tribunal_decision"]}, "tribunal_decision_category"=>{"any"=>["charity"]}}, links: {}, email_document_supertype: "", government_document_supertype: "", signon_user_uid: nil, slug: "upper-tribunal-tax-and-chancery-chamber-decisions-...", url: nil, tags_digest: "56ebfeb705da5037ebe24ed1617068a34810753649c60eb80b...", links_digest: nil, content_id: nil, description: nil, last_audited_at: nil, last_alerted_at: "2024-11-12 08:45:04.474051000 +0000"> ```

I then tried calling `find_or_create_subscriber_list` with the same links, but a new title:

``` finder-frontend(prod)> GdsApi.email_alert_api.find_or_create_subscriber_list({ "title" => "Upper Tribunal NEW", "tags" => { "format"=>{"any"=>["tax_tribunal_decision"]}, "tribunal_decision_category"=>{"any"=>["charity"]} } }) ```

We can see that `title` got updated on the existing subscriber list:

``` email-alert-api(prod)> existing.reload => id: 8330, title: "Upper Tribunal NEW", created_at: "2017-04-04 11:25:58.190324000 +0100", updated_at: "2025-01-14 10:45:26.063566000 +0000", document_type: "", tags: {"format"=>{"any"=>["tax_tribunal_decision"]}, "tribunal_decision_category"=>{"any"=>["charity"]}}, links: {}, ... ```

Therefore we can safely conclude that tweaking the subscriber list title generation logic won't have a detrimental effect on data fragmentation. But just to be sure, the team in charge of Email Alert API will monitor the effects of this change, in https://trello.com/c/5HrbIlhQ/287-confirm-status-of-email-lists-affected-by-pluralisation-simplification

## Next steps

- Remove the corresponding code from Finder Frontend: https://github.com/alphagov/finder-frontend/pull/3626
- Remove the hash option from the schema in Publishing API: TODO

---

⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Follow [these steps](https://guides.rubyonrails.org/upgrading_ruby_on_rails.html) if you are doing a Rails upgrade.

ChrisBAshton committed:
Remove pluralization of email subscription topic prefixes

Trello: https://trello.com/c/l9XfLavV/3212-email-alerts-on-edit-finder-form

Support for a hash value for `subscription_list_title_prefix` has
been around for ages (at least [9 years](https://github.com/alphagov/specialist-publisher/commit/b2e3d301804f1f5eddd040e8d4e3ecf50601a682#diff-793b8db55b8b5c75a5bf7007ed820f5038b1e7f6160a7f43a12477372fc3c3ceR16))
but it is only used by a handful of finders.

A [whole bunch of logic](https://github.com/alphagov/finder-frontend/blob/484396b660715811500af3e3b226ecf0dd69c675/app/lib/email_alert_title_builder.rb#L19-L47)
has to be maintained in finder-frontend to support plurality.
Until now, we haven't felt the pain of it in Specialist Publisher,
but we're now building a form whereby publishers can request edits
to the configuration of their finder. That means we either need to
add support for pluralised subscription topics in the form itself,
or skip over any existing hash values, or, as we're doing in this
commit, omit the hash values altogether. This commit argues that
the feature is poorly utilised, not very valuable, and is actively
blocking development of the "request edits to email alerts" form.

Only 6 finders made use of the feature, and for five of these,
we were able to trivially replace references to "...the following
{nouns|noun}" to "...the following noun(s)". There is one finder
that had "category" vs "categories", which is harder to substitute,
but [taking inspiration from other finders](https://github.com/alphagov/specialist-publisher/blob/b638d1326b570988d63aaec9de2bd6c2c877ea22/lib/documents/schemas/flood_and_coastal_erosion_risk_management_research_reports.json#L17)
we can simply follow the form "{Finder name} {document noun}",
with no need to overcomplicate things by trying to form a partial
sentence.

With this commit, we can delete the corresponding code from
finder-frontend.

NB, I wasn't sure what would happen to existing subscriber lists
if we were to go and change the subscriber list title (but
keeping all other params the same) - is a new one created, and if
so, is that even a problem? So I tried it out. Rails console on
Integration (Email Alert API), getting a subscriber list:

```
email-alert-api(prod)> SubscriberList.find_by(slug: "upper-tribunal-tax-and-chancery-chamber-decisions-with-the-following-category-charity")
=>
 id: 8330,
 title: "Upper Tribunal (Tax and Chancery Chamber) decision...",
 created_at: "2017-04-04 11:25:58.190324000 +0100",
 updated_at: "2019-09-16 13:26:39.390801000 +0100",
 document_type: "",
 tags: {"format"=>{"any"=>["tax_tribunal_decision"]}, "tribunal_decision_category"=>{"any"=>["charity"]}},
 links: {},
 email_document_supertype: "",
 government_document_supertype: "",
 signon_user_uid: nil,
 slug: "upper-tribunal-tax-and-chancery-chamber-decisions-...",
 url: nil,
 tags_digest: "56ebfeb705da5037ebe24ed1617068a34810753649c60eb80b...",
 links_digest: nil,
 content_id: nil,
 description: nil,
 last_audited_at: nil,
 last_alerted_at: "2024-11-12 08:45:04.474051000 +0000">
```

I then tried calling `find_or_create_subscriber_list` with the
same links, but a new title:

```
finder-frontend(prod)> GdsApi.email_alert_api.find_or_create_subscriber_list({ "title" => "Upper Tribunal NEW", "tags" => { "format"=>{"any"=>["tax_tribunal_decision"]}, "tribunal_decision_category"=>{"any"=>["charity"]} } })
```

We can see that `title` got updated on the existing subscriber list:

```
email-alert-api(prod)> existing.reload
=>
 id: 8330,
 title: "Upper Tribunal NEW",
 created_at: "2017-04-04 11:25:58.190324000 +0100",
 updated_at: "2025-01-14 10:45:26.063566000 +0000",
 document_type: "",
 tags: {"format"=>{"any"=>["tax_tribunal_decision"]}, "tribunal_decision_category"=>{"any"=>["charity"]}},
 links: {},
 ...
```

Therefore we can safely conclude that tweaking the subscriber list
title generation logic won't have a detrimental effect on data
fragmentation. But just to be sure, the team in charge of Email
Alert API will monitor the effects of this change, in
https://trello.com/c/5HrbIlhQ/287-confirm-status-of-email-lists-affected-by-pluralisation-simplification

KludgeKML commented on lib/documents/schemas/tax_tribunal_decisions.json:
```diff
@@ -22,10 +22,7 @@
-    "singular": "Upper Tribunal (Tax and Chancery Chamber) decisions with the following category: ",
-    "plural": "Upper Tribunal (Tax and Chancery Chamber) decisions with the following categories: "
-  },
+  "subscription_list_title_prefix": "Upper Tribunal (Tax and Chancery Chamber) decisions",
```
Would "**Upper Tribunal (Tax and Chancery Chamber) decisions categorised by:**" make sense? (probably best to ask a content designer)

ChrisBAshton replied:
Good prompt! I've now dug into the [corresponding logic in finder-frontend](https://github.com/alphagov/finder-frontend/blob/484396b660715811500af3e3b226ecf0dd69c675/app/lib/email_alert_title_builder.rb#L26). For strings, rather than hashes, the prefix is `"#{subscription_list_title_prefix.strip} with"`: so alongside the [suffix](https://github.com/alphagov/finder-frontend/blob/484396b660715811500af3e3b226ecf0dd69c675/app/lib/email_alert_title_builder.rb#L55) we'd have a value of "Upper Tribunal (Tax and Chancery Chamber) decisions with category Charity" (with what I currently have committed). I think that's ok.

(I.e. your suggestion would render as "Upper Tribunal (Tax and Chancery Chamber) decisions categorised by: with category Charity")

But that does mean I need to fix up my other updated values (e.g. we'd currently have "Algorithmic transparency records with the following function(s): with function General public services", which doesn't read great. I'll push up another commit.

KludgeKML replied:
Hmm, so has it always been awful in Finder-Frontend? Because the titles inside email-alert-api look okay, and I'm worried with this change (which includes removing a space) we'll go from (eg):

`CMA cases with the following case type: oim project`

..to..

`CMA casesoim project`

ChrisBAshton replied:
It shouldn't do. I've written some tests for finder-frontend in https://github.com/alphagov/finder-frontend/pull/3626/files and it looks as though the subject would come out as `CMA cases with case type oim project`

KludgeKML replied:
Ah, right - sorry, I was thinking this prefix was used by both FF and email alert api, but it's just used by FF and then the results are passed to email alert api. Cool.

ChrisBAshton committed:
Tweak subscription_list_title_prefix to work well with suffix

Off the back of a [code review comment](https://github.com/alphagov/specialist-publisher/pull/2946#discussion_r1914772071)
I've looked into how the prefix would work with what's currently
in the logic in finder-frontend.

The [corresponding logic in finder-frontend](https://github.com/alphagov/finder-frontend/blob/484396b660715811500af3e3b226ecf0dd69c675/app/lib/email_alert_title_builder.rb#L26)
has, for strings (rather than hashes), the prefix
`"#{subscription_list_title_prefix.strip} with"`. This means that
alongside the [suffix](https://github.com/alphagov/finder-frontend/blob/484396b660715811500af3e3b226ecf0dd69c675/app/lib/email_alert_title_builder.rb#L55)
we'd have a value of "Upper Tribunal (Tax and Chancery Chamber)
decisions with category Charity" (with what I currently have
already committed), and that looks ok.

Re-reviewing the other cases where I'd converted hashes to
sentences, it's clear we need a further tweak. For example,
without this commit, the algorithmic transparency records email
subscriber list would have a title of something like "Algorithmic
transparency records with the following function(s): with function
General public services", which does not make much sense! With
this commit, it would now read as "Algorithmic transparency
records with function General public services", which makes more
sense.

ryanb-gds reviewed (approved):
Nice to see this gone

KludgeKML reviewed (approved):
LGTM!
